In [12]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [13]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                   1.08388                     7.60278   
ethereum                                  0.22392                    10.38134   
tether                                   -0.21173                     0.04935   
ripple                                   -0.37819                    -0.60926   
bitcoin-cash                              2.90585                    17.09717   
binancecoin                               2.10423                    12.85511   
chainlink                                -0.23935                    20.69459   
cardano                                   0.00322                    13.99302   
litecoin                                 -0.06341                     6.60221   
bitcoin-cash-sv                           0.92530                     3.29641   
crypto-com-chain                          0.61209                    -5.67151   
usd-coin                                 -0.17825                    -0.11871   
eos                                       0.14477                    -1.31177   
monero                                    0.42961                    15.78515   
tron                                      0.07647                     4.23886   
tezos                                    -0.67316                     8.95665   
okb                                      -2.72700                    -4.55389   
stellar                                  -1.00843                     2.07149   
cosmos                                   -0.95103                    16.08534   
cdai                                      0.21169                     0.05820   
neo                                       0.49302                     2.44243   
wrapped-bitcoin                           1.10231                     7.40537   
leo-token                                -0.13192                    -1.34886   
huobi-token                              -0.40818                     1.61798   
nem                                      -0.84990                    -0.85140   
binance-usd                              -0.10642                     0.04726   
iota                                      0.41996                     6.06830   
vechain                                   1.28766                    -1.76352   
zcash                                    -0.60897                    10.34780   
theta-token                              -4.56089                    -6.09456   
dash                                     -1.06006                     5.09387   
ethereum-classic                         -0.45950                     3.05209   
ethlend                                 -13.52786                     4.21266   
maker                                    -0.60285                     8.15400   
havven                                   -4.07216                     4.33651   
omisego                                   4.84033                     6.82985   
celsius-degree-token                      2.51323                     0.60354   
ontology                                 -1.35845                    -1.21399   
ftx-token                                 0.83416                     7.10438   
true-usd                                 -0.06197                     0.16642   
digibyte                                  2.97084                     0.67038   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                   6.57509   
ethereum                                  4.80849   
tether                                    0.00640   
ripple                                    2.24984   
bitcoin-cash                             14.75334   
binancecoin                               6.80688   
chainlink                                 9.30098   
cardano                                   5.55476   
litecoin

In [14]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [15]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprec

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [16]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
crypto_scaled = StandardScaler().fit_transform(df_market_data)
crypto_scaled 

array([[ 5.08529366e-01,  4.93193071e-01,  7.72200433e-01,
         2.35459633e-01, -6.74950963e-02, -3.55953481e-01,
        -2.51636882e-01],
       [ 1.85445894e-01,  9.34445040e-01,  5.58692121e-01,
        -5.43409317e-02, -2.73482725e-01, -1.15759474e-01,
        -1.99352110e-01],
       [ 2.17739616e-02, -7.06336853e-01, -2.16804207e-02,
        -6.10301536e-02,  8.00452481e-03, -5.50246924e-01,
        -2.82060506e-01],
       [-4.07643829e-02, -8.10928066e-01,  2.49457974e-01,
        -5.03879651e-02, -3.73164019e-01, -4.58258816e-01,
        -2.95546142e-01],
       [ 1.19303608e+00,  2.00095907e+00,  1.76061001e+00,
         5.45842065e-01, -2.91202870e-01, -4.99847761e-01,
        -2.70316950e-01],
       [ 8.91870708e-01,  1.32729453e+00,  8.00214184e-01,
        -5.71478992e-02,  7.78653106e-01, -1.88231917e-01,
        -2.25532605e-01],
       [ 1.13972400e-02,  2.57225091e+00,  1.10164693e+00,
        -4.90495415e-01, -9.31954023e-01,  3.87758986e-01,
        -1.8284399

In [17]:
# Create a DataFrame with the scaled data
crypto_scale_df = pd.DataFrame(crypto_scaled, columns=['price_change_percentage_24h', 'price_change_percentage_7d', 'price_change_percentage_14d', 'price_change_percentage_30d', 'price_change_percentage_60d', 'price_change_percentage_200d', 'price_change_percentage_1y'])

# Copy the crypto names from the original data
crypto_names = list(df_market_data.index)

# Set the coinid column as index
crypto_scale_df['coin_id'] = crypto_names
crypto_scale_df = crypto_scale_df.set_index('coin_id')

# Display sample data
crypto_scale_df

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Data.

In [18]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [19]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    model = KMeans(n_clusters = i, random_state = 0)
    model.fit(crypto_scale_df)
    inertia.append(model.inertia_)

/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The defa

In [20]:
# Create a dictionary with the data to plot the Elbow curve
df_elbow_1 = {
    "k": k,
    "inertia" : inertia
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_1 = pd.DataFrame(df_elbow_1)
df_elbow_1

k     inertia
0   1  287.000000
1   2  198.571818
2   3  123.190482
3   4   79.022435
4   5   65.302379
5   6   52.101806
6   7   46.035250
7   8   37.284807
8   9   32.161417
9  10   28.075253

In [21]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow_1.hvplot.line(
    x = "k",
    y = "inertia",
    title = "Elbow Curve",
    xticks=k
)



/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/data/pandas.py:39: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  return dataset.data.dtypes[idx].type
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/data/pandas.py:39: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  return dataset.data.dtypes[idx].type


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [22]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters = 4, random_state = 0)


In [23]:
# Fit the K-Means model using the scaled data
model.fit(crypto_scale_df)


/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


KMeans(n_clusters=4, random_state=0)

In [24]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_4 = model.predict(crypto_scale_df)

# Print the resulting array of cluster values.
k_4

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

In [25]:
# Create a copy of the DataFrame
clusters_predicted = crypto_scale_df.copy()

In [26]:
# Add a new column to the DataFrame with the predicted clusters
clusters_predicted["CurrencyCluster"] = k_4 

# Display sample data
clusters_predicted

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [36]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
clusters_predicted.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    by = "CurrencyCluster",
    hover_cols = "coin_id"
    
)



/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/data/pandas.py:39: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, intege

:NdOverlay   [CurrencyCluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [37]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [38]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
currency_pca = pca.fit_transform(clusters_predicted)

# View the first five rows of the DataFrame. 
currency_pca[0:5]

array([[ 1.19408229, -0.90207402, -0.58533763],
       [ 1.00970605, -0.66358438, -1.13695032],
       [-0.72253591, -0.30763241,  0.7218133 ],
       [-0.74826646, -0.33237895,  0.55847869],
       [ 2.25853853, -1.82696614, -1.37816649]])

In [39]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_


array([0.37269822, 0.32489961, 0.18917649])

In [40]:
sum(pca.explained_variance_ratio_)


0.8867743104975919

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8867743104975919

In [41]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data
new_pca_df = pd.DataFrame(currency_pca, columns = ["PCA1","PCA2", "PCA3"])

# Copy the crypto names from the original data
new_pca_df["coin_id"] = df_market_data.index

# Set the coinid column as index
new_pca_df = new_pca_df.set_index("coin_id")

# Display sample data
new_pca_df

PCA1      PCA2      PCA3
coin_id                                           
bitcoin               1.194082 -0.902074 -0.585338
ethereum              1.009706 -0.663584 -1.136950
tether               -0.722536 -0.307632  0.721813
ripple               -0.748266 -0.332379  0.558479
bitcoin-cash          2.258539 -1.826966 -1.378166
binancecoin           1.719188 -0.993613 -0.663561
chainlink             1.396401 -0.721542 -2.731341
cardano               1.359187 -0.651114 -1.482162
litecoin              0.869062 -0.828724 -0.856779
bitcoin-cash-sv      -0.763186 -0.619275  0.315895
crypto-com-chain     -1.795991  0.250828  0.984707
usd-coin             -0.733773 -0.310013  0.737215
eos                  -1.024249 -0.466951  0.489504
monero                3.248514 -0.917857 -0.654640
tron                 -0.553164 -0.303202  0.213757
tezos                 0.159464 -0.702819 -1.624073
okb                  -1.559263  0.507701  0.744003
stellar              -1.093506 -0.206348  0.011414
cosmos                1.387824 -0.611337 -1.684985
cdai                 -0.727976 -0.391426  0.738362
neo                  -1.384449  0.007636  0.498712
wrapped-bitcoin       1.176562 -0.900333 -0.572220
leo-token            -1.201584 -0.126772  0.900321
huobi-token          -0.700054 -0.277189  0.459856
nem                   0.073392  0.479044  1.516378
binance-usd          -0.714467 -0.328951  0.731577
iota                 -0.946746 -0.540382 -0.163832
vechain              -2.297994  0.119851  0.569463
zcash                 1.066956 -0.645155 -1.212974
theta-token           0.064891  2.640044  2.132057
dash                 -0.853033 -0.411149 -0.245297
ethereum-classic     -0.820641 -0.404693  0.185554
ethlend              -0.321451  8.800190 -2.894631
maker                 0.694440 -0.496058 -0.759548
havven               -1.884981  1.533159 -0.749783
omisego              -0.177649  0.045948  1.457832
celsius-degree-token  7.610759  2.465053  3.719577
ontology             -2.361075  0.084567 -0.000994
ftx-token            -0.464345 -0.538034  0.054965
true-usd             -0.702148 -0.341378  0.726295
digibyte             -0.736439 -0.167072  0.929537

---

### Find the Best Value for k Using the PCA Data

In [42]:
# Create a list with the number of k-values from 1 to 11
k2 = list(range(1, 11))
k2

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [43]:
# Create an empty list to store the inertia values
inertia2 = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k2:
    model2 = KMeans(n_clusters=i, random_state=0)
    model2.fit(new_pca_df)
    inertia2.append(model2.inertia_)

/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The defa

In [44]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca_2 = {
    "k": k2,
    "inertia": inertia2
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca_2 = pd.DataFrame(elbow_data_pca_2)
df_elbow_pca_2


k     inertia
0   1  290.018457
1   2  199.108053
2   3  112.401201
3   4   43.586433
4   5   32.255267
5   6   23.915931
6   7   18.509832
7   8   15.653147
8   9   11.939564
9  10    9.189483

In [45]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow_pca_2.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve Using PCA Data", 
    xticks=k)



/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/data/pandas.py:39: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  return dataset.data.dtypes[idx].type
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/data/pandas.py:39: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  return dataset.data.dtypes[idx].type


:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it's the same value.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [46]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters = 4, random_state = 0)


In [47]:
# Fit the K-Means model using the PCA data
model.fit(new_pca_df)

/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


KMeans(n_clusters=4, random_state=0)

In [48]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_4 = model.predict(new_pca_df)
# Print the resulting array of cluster values.
k_4

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

In [49]:
# Create a copy of the DataFrame with the PCA data
copy_new_pca_df = new_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
copy_new_pca_df["CurrencyCluster"] = k_4

# Display sample data
copy_new_pca_df

PCA1      PCA2      PCA3  CurrencyCluster
coin_id                                                            
bitcoin               1.194082 -0.902074 -0.585338                1
ethereum              1.009706 -0.663584 -1.136950                1
tether               -0.722536 -0.307632  0.721813                0
ripple               -0.748266 -0.332379  0.558479                0
bitcoin-cash          2.258539 -1.826966 -1.378166                1
binancecoin           1.719188 -0.993613 -0.663561                1
chainlink             1.396401 -0.721542 -2.731341                1
cardano               1.359187 -0.651114 -1.482162                1
litecoin              0.869062 -0.828724 -0.856779                1
bitcoin-cash-sv      -0.763186 -0.619275  0.315895                0
crypto-com-chain     -1.795991  0.250828  0.984707                0
usd-coin             -0.733773 -0.310013  0.737215                0
eos                  -1.024249 -0.466951  0.489504                0
monero                3.248514 -0.917857 -0.654640                1
tron                 -0.553164 -0.303202  0.213757                0
tezos                 0.159464 -0.702819 -1.624073                1
okb                  -1.559263  0.507701  0.744003                0
stellar              -1.093506 -0.206348  0.011414                0
cosmos                1.387824 -0.611337 -1.684985                1
cdai                 -0.727976 -0.391426  0.738362                0
neo                  -1.384449  0.007636  0.498712                0
wrapped-bitcoin       1.176562 -0.900333 -0.572220                1
leo-token            -1.201584 -0.126772  0.900321                0
huobi-token          -0.700054 -0.277189  0.459856                0
nem                   0.073392  0.479044  1.516378                0
binance-usd          -0.714467 -0.328951  0.731577                0
iota                 -0.946746 -0.540382 -0.163832                0
vechain              -2.297994  0.119851  0.569463                0
zcash                 1.066956 -0.645155 -1.212974                1
theta-token           0.064891  2.640044  2.132057                0
dash                 -0.853033 -0.411149 -0.245297                0
ethereum-classic     -0.820641 -0.404693  0.185554                0
ethlend              -0.321451  8.800190 -2.894631                2
maker                 0.694440 -0.496058 -0.759548                1
havven               -1.884981  1.533159 -0.749783                0
omisego              -0.177649  0.045948  1.457832                0
celsius-degree-token  7.610759  2.465053  3.719577                3
ontology             -2.361075  0.084567 -0.000994                0
ftx-token            -0.464345 -0.538034  0.054965                0
true-usd             -0.702148 -0.341378  0.726295                0
digibyte             -0.736439 -0.167072  0.929537                0

In [50]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
copy_new_pca_df.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="CurrencyCluster",
    hover_cols = ["coin_id"], 
)

/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/data/pandas.py:39: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, intege

:NdOverlay   [CurrencyCluster]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [56]:
# Composite plot to contrast the Elbow curves
df_elbow_1.hvplot.line(x="k", y="inertia", title= "Elbow Curve 1", xticks=k) + df_elbow_pca_2.hvplot.line(x="k", y="inertia", title="Elbow Curve Using PCA Data", xticks=k)



/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/data/pandas.py:39: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  return dataset.data.dtypes[idx].type
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/data/pandas.py:39: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  return dataset.data.dtypes[idx].type
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/data/pandas.py:39: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame 

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [54]:
# Composite plot to contrast the clusters
clusters_predicted.hvplot.scatter( x = "price_change_percentage_24h", y = "price_change_percentage_7d", by = "CurrencyCluster", hover_cols = "coin_id") + copy_new_pca_df.hvplot.scatter(x="PCA1", y="PCA2", by="CurrencyCluster", hover_cols = ["coin_id"])

/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  return pd.unique(values)
/Users/evermz/anaconda3/envs/ever/lib/python3.9/site-packages/holoviews/core/util.py:1175: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprec

:Layout
   .NdOverlay.I  :NdOverlay   [CurrencyCluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [CurrencyCluster]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** It's easier to understand and interpret the plot with PCA. This is because we have tighter clusters and defined. Also, using PCA we were able to eliminate some outliers, as we can see in the graphs above.